engine/src/mblandroidjava.cpp

```diff
@@ -439,6 +453,7 @@ bool MCJavaInitList(JNIEnv *env, jobject &r_list)
 bool MCJavaFreeList(JNIEnv *env, jobject p_list)
 {
     env->DeleteLocalRef(p_list);
+    free_arraylist_class(env);
```

Comment:
You don't want to free the arraylist class here - we want to keep the globalref around so we don't have to fetch it each time we create a list.

Whilst we are fixing this issue we should really make the statics in this file work correctly - Android does not reinit statics on restart of an app which means they can contain stale data.

We need a MCJavaInitialize / MCJavaFinalize method in this file - Initialize should call all the 'class' init functions (init_*_class), and Finalize the 'free_*_class' functions). The Initialize/Finalize pair should be added to the android engine startup (in mobile_main in mblandroiddc.cpp - after we attach to the thread).